docs/getting-started/nullblk.md

```diff
@@ -1,23 +1,28 @@
 # Zoned Block Device Emulation with *null_blk*
 
-Linux&reg; *null_blk* driver is a powerful tool to emulate various types of
-block devices for tests. Since kernel 4.19, the *null_blk* driver gain the
-ability to emulate zoned block devices. With the addition of memory backup for
-data read and writen to a *null_blk* device, this driver provides an easy to use
-but yet powerful tool for application development and tests.
+The 
+[Linux&reg; *null_blk* driver](https://www.kernel.org/doc/Documentation/block/null_blk.txt)
+is a powerful tool that can emulate various types of block devices.
+Since kernel version 4.19, the *null_blk* driver has been able to
+emulate zoned block devices. Because memory backup has been added to
+the *null_blk* device for data reading and writing operations, the
+*null_blk* driver has become a practical, powerful tool that can be
+used for application development and tests.
 
 ## Creating a Zoned *null* Block Device
 
-The simplest method to create a *null_blk* emulated zoned block device is to
-specify the `zoned=1` argument to *null_blk* modrpobe command line.
+The simplest method for creating a *null_blk* emulated zoned block
+device is to specify `zoned=1` as an argument that follows the
+*modprobe **null_blk* command on the command line, as in the following
+example:
 
 ```plaintext
 # modprobe null_blk nr_devices=1 zoned=1
 ```
 
-This creates a single host managed zoned block device with a zone size of 256M
-and a total capacity of 250 GB (1000 zones). No conventional zones are created
-with this simple command.
+This creates a single-host-managed zoned block device with a zone size
```

Comment:
Why add the hyphens in "single-host-managed" ?

Reply:
**Quick Answer**
We need hyphens in English in order to form compound adjectives. 

**Not-so Quick Answer**
We need these hyphens and these compound adjectives because English in the 21st century is so overwhelmingly analytic (as opposed to synthetic, like Latin or Russian or German) that it doesn't have adjective endings *(see the note at the end of this section). That means that only context and experience can be expected to lead the reader to the proper interpretation of the sentence or clause under study. In this case, the clause "This creates a single host managed zoned block device with a zone size" is hard to interpret without some orthographic signposting in the form of these hyphenated complex adjectives. And that is a serious problem when you're writing instructional material, because instructional material is intended to provide precisely the kind of context and experience that is necessary to interpret these kinds of clauses.

Pages 1656-1660 of the _Cambridge Grammar of the English Language_ provide a thorough (and maybe even irritatingly hair-splitting) account of compound adjectives. 

What I said earlier about English not having adjective endings isn't entirely true. English has a single adjective that takes a gender: "blond" when a male's hair is yellow-colored, and "blonde" when a female's hair is yellow-colored.

**Questions For You That Follow From Your Question To Me**

As I read this again, I see that I might have misinterpreted the sentence. So, here are two versions of the sentence. Tell me which one is correct.

    1. "This creates a single, host-managed zoned block device with a zone size..."

    In this version of this clause, a single zoned block device is created, and it is managed by a host.

    2. "This creates a single-host-managed zoned block device with a zone size..."

    In this version of this clause, emphasis is put on the fact that the zoned block device is managed by a single host.
     Candidly, I'll tell you that on re-reading this, I suspect that this is incorrect and that version 1 above is correct.

**Finally**
I could go on about the ways that we have to kluge our way around the deficiencies in English (for example, we have no antonym for the verb "to exceed", a lack that results in formations like "make sure that the level of water does not exceed" ), but I suspect that I have beaten a horse that, if not currently dead, is in mortal danger of expiring.

Reply:
I prefer "This creates a single, host-managed zoned block device with a zone size..." (version 1). But I am heavily biased, and no matter how you write it, I still get the meaning correctly :)
Of note is that the sysfs "zoned attribute file gives "host-managed" with a hyphen. That was to avoid a space, which causes problems with functions like fscanf() to get the file value as a string. BUT, T10 and T13 standards do not use hyphens and refer to the models as "host managed", "host aware" and "drive managed". Whenever possible, I try to use the same wording as the standards to avoid confusions. But that can come at the cost of bad grammar as in this case...

Reply:
I agree with you in the matter of choosing between Sentence 1 and Sentence 2.

It is no surprise that the meaning of the text is clear to you. Your exquisite degree of expertise furnishes your mind with the objects under discussion, and keeps your mind on them without fail even when encountering language that refers to them in an imprecise way that might confuse someone whose only awareness of these objects has been furnished by that text.

I agree with you that text should adhere to standards (for example, T10 and T13 standards) whenever possible.

Reply:
This note isn't directed toward Damien. I put it here for posterity, because this sentence is a good example of a sentence that is difficult to parse if you don't already know what it refers to, and I've long entertained the ambition of writing about such a sentence.

**The Sentence In Question**
> This creates a single host managed zoned block device with a zone size of 256M
> and a total capacity of 250 GB (1000 zones).

**What The Sentence Means**

- "This" (the first word in the sentence) anaphorically refers to a modprobe command, namely "# modprobe null_blk nr_devices=1 zoned=1".
- The modprobe command creates a lone "zoned block device". That is, the modprobe command creates one and only one "zoned block device".
- The "zoned block device" that is created by this modprobe command has a zone size of 256M.
- The "zoned block device" that is created by this modprobe command has a total capacity of 250 GB (1000 zones).

**Why This Sentence Is Hard To Interpret**

1. The phrase "zoned block device" is a complex noun consisting of three words. The first word ("zoned") is an adjective and unambiguously an adjective (I think a case could be made here for "zoned" being a present participle functioning in this context as an adjective, but this is in the present sentence too much of a tangent even for me), but the second word ("block") seems to be a noun or perhaps an adjectival noun (like "prison" in the phrase "prison notebooks"... where the noun "prison" is pushed beyond the limits of its part of speech into service as an adjective), and the third word ("device") is unambiguously a noun. So the naive reader is facing a trebled challenge when she tries to interpret the phrase "zoned block device".
2. The adjectives "single" and "host managed", when presented serially with no signposting orthographic aid, can be interpreted by the naive reader (as described in one of my comments above) to mean either "managed by a single host" (single-host managed) or "one and only one block device that is managed by the host (single, host-managed). I feel confident that other interpretations are possible, but I think that the ambiguity here is sufficiently demonstrated.
3. "a single host managed zoned block device" is, then, just one hell of a direct object for the naive reader to interpret. Here we have a noun phrase of seven words that consists of what looks to the naive reader like an article (a), a noun (host), a past participle (managed), another past participle (zoned), a noun (block), and a noun (device).
4. The prepositional phrase "with a zone size of 256M and a total capacity of 250 GB (1000 zones)" could possibly be interpreted to be instrumental because in English, with-clauses and with-phrases are sometimes instrumental (cf. The Simpsons, S09E19, "Simpson Tide": "So, what's the best way to get peace?" "With a knife!"). Now, to the seasoned veteran of zoned storage, it is quite obvious that the "zone size" is not the instrument that the modprobe command employs in the act of creating the zoned block device. But in the mind of the naive reader, the presence of the preposition "with" suggests the possibility that such an instrumental clause has been deployed, and such an interpretation cannot be ruled out until the whole sentence has been read and the mind of the reader has slotted each part of the sentence into its proper syntactic role. Anything that can be done to make it easier for the reader to understand the proper syntactic roles of each part of the sentence must be done.

**What I'm Going To Do About It**

I'm going to rewrite the sentence so that it looks like this:

> This creates a single, host-managed zoned block device that has a zone
> size of 256M and a total capacity of 250 GB (1000 zones).

**Zac, Are You Going To Nitpick Every Sentence On zonedstorage.io Like This?**

Good God, no.